### DISCUSSION TITLE
Modernize PixiJS in v7

### DISCUSSION THREAD

**Discussion by bigtimebuddy** [pixijs/pixijs#7541] [RFC]
# 📣  Overview

I think with IE sunsetting support next summer and it being 2021 and browsers are way better than when we started, it's time to reconsider a few aspects of PixiJS and how we ship the library. Here's a summary of the changes I'd like to see for the next major version:

## 🚀  Built for ES6+

Currently, PixiJS is build for ES5+ browsers. Practically this means IE 9-11 and some very old Android devices. Would like to remove the transpiling we do with Buble and ship TypeScript's ES6 target or (esbuild). There are a few benefits: faster build times, marginally smaller files size (class vs prototype chain), better run-time enforcement of scope (var vs const), better features (String's `startsWith`, Array's `reduce`, etc). For users that require the old browsers, they can use v6 or below or transpile themselves.

##  🗑️   Remove Polyfills

Similar to ES6, we are polyfilling a few APIs. These should be removed as they are all now standard and available in the majority of browsers. Users that have novel browser support, should polyfill themselves. Here's the rough list of polyfilled APIs:

* Math.sign
* Number.isInteger
* Object.assign
* Promise
* requestAnimationFrame
* Float32Array
* Uint32Array
* Uint16Array
* Uint8Array
* Int32Array
* ArrayBuffer

## 👋  Drop Canvas Support

At PixiJS's beginning WebGL did not have wide-spread support. It was in the ~50% marketshare so dual rendering WebGL/Canvas was more critical then. However, now [97%+](https://caniuse.com/?search=webgl) have support for WebGL. Also, looking into the future, the core team is considering WebGPU as a faster alternative to WebGL. At that point, we do not want to continue to support CanvasRenderer (Context2D rendering to be more specific) and WebGL & WebGPU. Continuing to support canvas rendering adds a tremendous amount of maintenance and complexity. Removing this would allow us to focus on the future of 2D rendering that's faster and more fully featured. For users that still need canvas, there still v3-6 that users can use, those aren't disappearing.

By removing CanvasRenderer we'd be able to remove AbstractRenderer and autoDetectRenderer and isWebGLSupported APIs, to name a few. As well as the _renderCanvas implementation for various DisplayObjects. Much cleaner!

**Comment by ivanpopelyshev** [comment]
>  Drop Canvas Support

* `generateCanvasTexture` is one of a heck of popular function
* it helps to test things if someone breaks webgl render function, or whether our Graphics implementation is good enough
* good ssr with node canvas, people even want us to add skia renderer in the list
* v5-v6 support of canvas is much more serious than v4
* we'll even get new features when new 3d matrices will hit! pixi-projection will have even better canvas support!

I also remind that even the biggest 2d webgl app https://crazypanda.ru/en/game/the-household/ cant survive without CanvasRenderer. Yes, it works on a fork and wont be affected by any patches here, but the point stays - 2d is useful.

**Reply by bigtimebuddy** [reply]
As I was writing this I thought, "Ivan is going to the first to argue about dropping canvas". I see your points but I still don't agree keeping it is a good strategy long term.

- nothing is stopping people from using canvas either using another library, using context2d directly or continuing to use v6, this is about the _future_
- I consider CanvasRenderer feature-complete, I do not want to support new context2d features and would firmly disagree with adding
- likely we can support v6 for any bug fixes related to canvas for awhile, we are still fixing critical things in v5
- SSR isn't a think yet for us (tho some have hacked it) and we should be looking at WebGL rendering SSR not canvas when we officially support
- in general I think you're over valuing canvas and undervaluing the maintenance complexity imposed by having it in the first place

I'm still firm in this not being in the library. But I could definitely see another pixi-spine-like or filters-like project that owns the problem area. For the majority of cases, user are doing webgl only. Npm says weekly download of pixi.js is 45k and pixi.js-legacy is 15k. I suspect this will continue to change over time.

**Reply by domis86** [reply]
> nothing is stopping people from using canvas either using another library, using context2d directly or continuing to use v6, this is about the future

I just want to point out that canvas support (or "fallback") is one of most known Pixi.js core features - and it shows in descriptions of Pixi.js across the web - examples:

* https://www.javascripting.com/view/pixi-js - `Super fast HTML 5 2D rendering engine that uses webGL with canvas fallback ... PixiJS has full WebGL support and seamlessly falls back to HTML5's canvas if needed. `
* https://fr.wikipedia.org/wiki/Pixi.js - `Elle utilise la technologie WebGL par défaut et comporte un fallback dans la technologie HTML5 canvas`
* https://gist.github.com/dmnsgn/76878ba6903cf15789b712464875cfdc - `PixiJS: Super fast HTML 5 2D rendering engine that uses webGL with canvas fallback`
* https://blog.logrocket.com/top-6-javascript-and-html5-game-engines/ - `Pros: ... Support for WebGL and Canvas fallback: Pixi.js uses a WebGL renderer but also supports a Canvas fallback`
* https://ourcodeworld.com/articles/read/308/top-15-best-open-source-javascript-game-engines - `Pixi.js is a super fast HTML 5 2D rendering engine that uses webGL with canvas fallback.  ... Pixi.js has full WebGL support and seamlessly falls back to HTML5's canvas if needed. ...`
* https://www.slant.co/versus/1965/5128/~pixi-js_vs_unreal-engine-4 - here Pixi.js has 2 "PROS" points one of which is: ` Supports WebGL w/ canvas fallback - Pixi is a WebGL renderer, but can fall back to canvas if WebGL is not supported or turned off.`

So if you remove this core feature then those descriptions will need to change probably. 

> undervaluing the maintenance complexity imposed by having it in the first place

That is other side of the coin. Maybe you are right that this feature is not widely used and removing it would ease development of other things. But again - it is always associated with Pixi.js wherever Pixi.js is mentioned. So i would suggest to carefuly think again about removing it :)

**Reply by ivanpopelyshev** [reply]
One extra big reason: **multiple contexts** . If we dont implement hack with one big context and many ImageBitmaps, we cant support it without canvas2d

**Reply by GoodBoyDigital** [reply]
> One extra big reason: **multiple contexts** . If we dont implement hack with one big context and many ImageBitmaps, we cant support it without canvas2d

fwiw - WebGPU has one context for all canvas elements too!

**Reply by ivanpopelyshev** [reply]
Btw does someone have example with one webgl context and several ImageBitmapped contexts?

**Comment by ivanpopelyshev** [comment]
More canvas 2d
* Good learning ground for new contributors
* Non premultiplied blending modes, we have only part of them through `@pixi/picture`

Yes, for special operations users can use pure 2d API and then do `Texture.from(canvas)` without our canvas modules.

**Comment by GoodBoyDigital** [comment]
Heya! Thanks for getting that thought out there @bigtimebuddy ! Been noodling on this for a while myself..

When we made Pixi back in the day, it was only chrome and Firefox that supported WebGL. At that time it was essential that Canvas fall back was required to reach 100% of devices. Fast forward 7(!!) years and the landscape has changed a fair bit! WebGL is supported by pretty much any browser that supports canvas these days. Overall the trend for Canvas usage seems to be that less people are depending on it. Making it something less 'core' as time goes by.

There is also new shiny new WebGPU tech on the horizon that will be landing in most major browsers in the next couple of years (including mobile). I am keen to invest time and energy here, so that when the time comes the majority will get a massive speed boost :D

Looking after the users that want and still need the canvas renderer is still essential of corse, this can be managed in a few ways:

1 -  (meh..) They can use any existing current version of pixi to have a canvas renderer
2 - (awesome) We can still have a canvas renderer, but as a separate plugin. Just like Spine and all the other awesome stuff that has been create. 

Appreciate an extra step for devs that want to use canvas renderer is not the best for all, every decision like this is bound to ruffle a few feathers. Not denying its usefulness at all,  but the complexity / size / time savings we would get mean that the vast majority of developers will get a better PixiJS experience. 

Having said that, would be great to continue this conversation and get a better understanding of the use cases we may not have considered and that might not work using the above two solutions.

**Comment by bigtimebuddy** [comment]
I've created a next branch as a proof-of-concept that delivers on all these promises above.

## 🔗  Links

Branch: https://github.com/pixijs/pixijs/tree/next
Build: https://pixijs.download/next/pixi.min.js
Docs: https://pixijs.download/next/docs/index.html

### Changes

* Remove all canvas packages
* Remove pixi.js-legacy bundle
* Replace InteractionManager with EventsSystem
* Remove AbstractRenderer
* Remove polyfills package
* Build as ES6, remove transpiling step

https://github.com/pixijs/pixijs/compare/dev...next

### TODOs

* AccessibilityManager is current broken because of events

### Benefits

* Output code is 10% smaller without transpiling/es5 overhead
* Faster on bunny-mark: ~14-15fps v6 vs ~18-19fps next
* Way simpler code-base, removes canvas hacks

**Reply by GoodBoyDigital** [reply]
Wow thats almost 30% faster!!!

**Reply by bigtimebuddy** [reply]
Probably most important to me: if you look at the compare link, this removes about ~13K lines of code from the project. For such a large project with a huge API surface, dramatically shrinking it means we can focus on the APIs we already have.

**Reply by themoonrat** [reply]
For such a performance improvement.... it makes me wonder what it is exactly that's giving such a huge performance benefit? Is there some low hanging fruit to investigate for v6?

**Reply by bigtimebuddy** [reply]
Was just chatting about this with @GoodBoyDigital.  I think first I need to do a broader test other than macOS-latest-Chrome to understand if this is a consistent performance improvement for all browsers.  I have hunches (maybe optimal with less scope-leak without any `var`s, less transpiling overhead, use of `class` instead of prototype chains). I dunno, anyone have other ideas? 

There are certainly things I could try in v6, like turning off transpiling (but with TS target es5) and see if that shows the same benefits. Good suggestion @themoonrat.

Update: I tested Safari and Firefox on macOS and Safari on iOS all of these have no difference performance-wise. I am still seeing 30% improvement in Chrome.

**Comment by miltoncandelero** [comment]
I support the idea of having the user polyfill if they need to and tagetting es6 is a good step forward.
~~we are also polyfilling the `eval` with the `unsafeEval` package.~~ What's the stance on that? 🤔 
(I just checked, it's not an `eval` polyfill but a `new Function` ponyfill, still, what is the stance?)

**Reply by ivanpopelyshev** [reply]
we cant remove unsafeEval because of some reasons, including this long discussion we had a month ago or so...

**Reply by bigtimebuddy** [reply]
This is a little off-topic. 

unsafe-eval is required for _some_ environments that have a stricter CSP, which isn't everywhere or for everyone. This optional package comes at a slight performance cost, so it is recommended only in cases where it's necessary (e.g., publishing requirements for 3rd party platforms).

I have a minor TODO for unsafe-eval. @qtiki pointed out in #7324 that we shouldn't do a try/catch to check if `new Function` to check if it is supported. Still planning on fixing that, but no plans to change the default behavior with regards to unsafe-eval.

**Comment by themoonrat** [comment]
**Built for ES6+**
"For users that require the old browsers, they can use v6 or below or transpile themselves.".
Could we supply an npx command that transpiles a build to es5? Or a repo that 'es5-ifys' a pixi v7 build? Trying to think of a path of least resistance to help those with this requirement.

**Remove Polyfills**
We can easily create a url via https://polyfill.io/ that would list the required polyfills. For example, would a plain https://polyfill.io/v3/polyfill.min.js?features=es6 work, or would there be others that need explicitly adding? 

**Drop Canvas Support**
This one is interesting; you list that 97% of users _can_ support WebGL, but then you take out those with performance caveats, those without stencil support, and you're at a lower figure. And I often forced devices with issues that seemed to be GPU driver related onto the Canvas renderer too. It was nice having that option to manually these devices onto that renderer, since Canvas always worked everywhere.

But I get that there's a lot of code using this. How much effort to 'plugin-ify' canvas support as a separate package? This the kind of work worth a bounty for?


If the above support networks aren't possible, then perhaps a strong statement of commitment is required that v6 will carry on receiving bug fixes (but not new features) for at least x period of time. I know we already do that with v5 now we're developing v6 if something is critical, but developers on v4 or v5 still are more about keeping things stable, it's their choice not to upgrade. But v6 to v7 may make it impossible for projects to ever be updated; and that'd be on us, not the devs.

**Reply by themoonrat** [reply]
PS I'm broadly in support of this modernization! "x% speed increase by dropping legacy" is a very clear reason to move ahead with this that everyone would understand. Just want us to be as clear and as helpful as we can on the upgrade path and support of existing users

**Reply by bigtimebuddy** [reply]
About stencil support... That is one that I'd like to change. Using stencil should probably not be a _requirement_ for PixiJS on the whole, just for using stencils (Graphics as mask). I'd be in favor of removing this requirement up-front for detection and only warn if users try to use stencils.

**Reply by bigtimebuddy** [reply]
You ask some good questions here and provide a lot of options. Personally, I'd be in favor of the community owning more of these canvas-problems and we can provide space to amplify those solutions (maybe currently how people render pixi using node.js and we point users to those options). A plugin option. hosted on pixijs' org, that is down-stream from core pixi seems like a natural way for that to evolve if at all. Also, explicit LTS (long-term support) for v6 is certainly an easy-ish compromise for say, the next two years.

**Reply by miltoncandelero** [reply]
About the polyfills I checked what we are including (which is not exactly the same that we say on the README)

https://polyfill.io/v3/polyfill.min.js?features=Math.sign%2CNumber.isInteger%2CObject.assign%2CPromise%2CArrayBuffer%2CFloat32Array%2CUint32Array%2CUint16Array%2CUint8Array%2CInt32Array%2CDate.now%2Cperformance.now%2CrequestAnimationFrame

we have:
- Math.sign
- Math.isInteger
- Object.assign
- Promise
- Date.now
- performance.now
- Typed arrays (UInt, Int, Float and Buffer)
- requestAnimationFrame

**Comment by manju-reddys** [comment]
Any consideration would be given to improve the 2D graphic's quality without loosing the performance? Like drawing line, circle etc.. In the V6 the quality of these elements are not good as zoomed, they look more of triangle than smooth.

See https://github.com/pixijs/pixijs/discussions/7372

**Reply by ivanpopelyshev** [reply]
Surprise!

https://github.com/pixijs/graphics-smooth

**Comment by jbcz-tv** [comment]
>  Drop Canvas Support

We are **strongly** against this. Our project uses PIXI as the rendering engine on a large scale, before we replace `pixi` it with `pixi-legacy` , a lot of users complained about black screen problem caused by performance carvet, hardware / driver problems. I can still vividly remember the miserable scene of the customer service channels being jammed with such problems when the product went live. We couldn't abandon this group of users.

There are still a lot of devices around the world that don't run WebGL properly due to graphics driver bugs or hardware issues, including the machine I'm using for development (actually, I purchased it last year, a brand-new machine with GTX 1650 Super). We shouldn't only consider browser compatibility, more factors should be taken into account.

**Reply by bigtimebuddy** [reply]
We have had time to think more critically about CanvasRenderer, and we decided we will not be dropping it after all in v7. There are a lot of legitimate use-cases for canvas-only rendering after hearing more feedback like this and comments above.

We, however, have some ideas how we can improve the CanvasRenderer:

* We are going to refactor the CanvasRenderer to share the composable _systems_ architecture with Renderer so that we can reuse more internals for things like managing the canvas, resizing, etc.
* We are considering creating a canvas-only PixiJS bundle, the benefit is that it would be a _very small_ food-print. We think it's possible and there are some legit use-case for getting PixiJS API, but a minimal as possible.
* We will either rename `pixi.js-legacy` or merge canvas fallback into `pixi.js`, still TBD

**Reply by Ayfri** [reply]
I investigated a bit about ES6 and found out that ES7 is also pretty old and we'll supported (more exactly the features that comes with ES7, as browsers don't follow ES versions). Could we use ES7 to have `array.includes()` and exponential operator `**` _yup that's all_ ?

Or maybe even ES8 with `string.padStart/End`, `object.getOwnPropertyDescriptors`, Atomics and shared memory (but browser support for this is not aligned with other ES8 features, so maybe compile to ES8 but wait a few years before using Atomics), trailing commas, `Object.values/entries` and most importantly, asynchronous functions.
But maybe because of Atomics ES8 support is far from being considered ? We can note that asynchronous functions could change a lot of the APIs and architectures used. As you could use a lot more Promises than now without having callback hells everywhere (I mean for people who can't use `async` syntax).

We could even consider ES9 but it's maybe still too recent 👀

**Reply by domis86** [reply]
@bigtimebuddy  nice!

**Comment by Julien-Marcou** [comment]
@Ayfri, I was thinking the same, I added a discussion about targeting newer ECMAScript version for PixiJS v7: https://github.com/pixijs/pixijs/discussions/8454
